compiler/optimizer/VPHandlers.cpp

```diff
@@ -855,6 +863,23 @@ static bool refineUnsafeAccess(OMR::ValuePropagation *vp, TR::Node *node)
    if (i >= layout->count())
       return refuseToConstrainUnsafe(vp, node, "no matching field");
 
+   if (field->_isVolatile != sym->isVolatile())
+      {
+      // We don't have a representation for field shadows with the unexpected
+      // volatility, and it's not straightforward to add one because we'd have
+      // to move volatile status to SymbolReference from Symbol.
+      //
+      // Functionally it would be fine to go on and refine anyway if the field
+      // is volatile. Doing so would still improve aliasing, but it would make
+      // the access itself more expensive, so just leave it alone for now.
+      //
```

Comment:
If we had a non-volatile access to a volatile field, maybe this is the case that you are saying the access would become more expensive (though functionally correct). 

If we had a volatile access to a non-volatile field, maybe this is the case that we would have a functional bug if we simply converted that to a non-volatile refined symbol.

Reply:
> in the case when it was volatile before the transformation and volatile after it as well ?
In that case we don't enter here because both the unsafe shadow and the refined shadow are volatile.

> If we had a non-volatile access to a volatile field, maybe this is the case that you are saying the access would become more expensive (though functionally correct).

Yes, effectively we would replace the non-volatile access with a non-volatile one.

> If we had a volatile access to a non-volatile field, this is the case that we would have a functional bug if we simply converted that to a non-volatile refined symbol. I am guessing this is the actual bug you encountered that forced the revert.

Right. This code (without the volatile handling) would ignore volatile completely, so it could replace a volatile unsafe shadow (volatile access) with a non-volatile array or fabricated field shadow (non-volatile access).

> Maybe your point is that you don't want to do the latter transformation (since it is wrong) and you also don't want to do the former transformation since it will be more expensive to access at run time anyway.

Precisely.

> I am assuming we handle the case when we had a volatile access to a volatile field in the transformation correctly, i.e. it would get transformed with the implementation and there is no functional issue with that.

This fix shouldn't interfere with the transformation in that case, since we'll have `field->_isVolatile != sym->isVolatile()`. I don't know of a functional issue with doing that or any particular reason to expect one